prow/inrepoconfig/api/api.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package api
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"github.com/sirupsen/logrus"
+	"sigs.k8s.io/yaml"
+
+	checkconfigapi "k8s.io/test-infra/prow/cmd/checkconfig/api"
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/git"
+	"k8s.io/test-infra/prow/github"
+)
+
+const (
+	// PluginName is the name of this plugin
+	PluginName = "inrepoconfig"
+
+	// ContextName is the name of the context the inrepoconfig plugin creates at GitHub
+	ContextName = "prow-config-parsing"
+
+	// ConfigFileName is the name of the configfile the inrepoconfig plugin uses to read its
+	// config from
+	ConfigFileName = "prow.yaml"
+)
+
+// InRepoConfig is the type `prow.yaml` gets deserialized into
+type InRepoConfig struct {
+	Presubmits []config.Presubmit `json:"presubmits,omitempty"`
+}
+
+// defaultAndValidateInRepoConfig defaults and validates the InRepoConfig. This must be called before
+// accessing any data in it
+func (irc *InRepoConfig) defaultAndValidateInRepoConfig(c *config.Config, orgRepo string) error {
+	for i := range irc.Presubmits {
+		ps := &irc.Presubmits[i]
+		config.DefaultPresubmitFields(&c.ProwConfig, ps)
+		if ps.Decorate {
+			ps.DecorationConfig = ps.DecorationConfig.ApplyDefault(c.Plank.DefaultDecorationConfig)
+		}
+		if err := config.ResolvePresets(
+			ps.Name, ps.Labels, ps.Spec, ps.BuildSpec, c.Presets); err != nil {
+			return err
+		}
+	}
+
+	// Has to be done in the end, otherwise the jobs Trigger field is unset
+	if err := config.SetPresubmitRegexes(irc.Presubmits); err != nil {
```

Comment:
Shouldn't this happen from load? Might want to DRY that out so you don't need to call out specifically to presets, regex,  etc

Reply:
Yeah as mentioned here https://github.com/kubernetes/test-infra/pull/12836/files#diff-21c85d95fdbaa35a9424bb69d4ab3379R79 the whole config defaulting and validation is not very well accessible from the outside right now, to improve that situation, a bigger refactor of `prow/config` will be needed. If anyhow possible I would really like to do that in a follow-up thought.

Reply:
Right now there are tests that test every single of those parts, so I hope this is good enough for a first version